### PR TITLE
Added the code to calculate factorial of number recursively

### PR DESCRIPTION
**Pull request by ambicadeshpande**
* Added the code to calculate the factorial of number using recursive function call
* Scenarios covered [UnitTest] 
    Factorial for number > 20 - long overflow
    Factorial for number < 0 - long overflow
    Factorial for any number (1-20) 
    Factorial for number = 0

### CODE REVIEW

File: Factorial/src/Factorial.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        System.out.println("factorialForAnyNumTest - passed");
+    }
+
+    @Test
+    public void factorialForZeroTest() {
+        long numZero = 0;
+        long actual = factorialUsingRecursion(numZero);
+        long expected = 1;
+        assertEquals(actual, expected);
+        System.out.println("factorialForZeroTest - passed");
```
Comment (rohmulan) [inline Factorial/src/Factorial.java]: There is no need to print this out

File: Factorial/src/Factorial.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import org.junit.jupiter.api.Test;
+import org.junit.runner.JUnitCore;
+import org.junit.runner.Result;
+import org.junit.runner.notification.Failure;
+import static org.junit.jupiter.api.Assertions.*;
+
+/* Reads an integer and prints n! = 1 * 2 * ... * n to standard output
+   For eg: 0! = 1, 1! = 1, 2! = 2 * 1 = 2, 3! = 3 * 2 * 1 = 6 ..  so on
```
Comment (rohmulan) [inline Factorial/src/Factorial.java]: Why comments? Your unit test cases are documentation enough

File: Factorial/src/Factorial.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+   Remarks
+    - There would be an overflow for long if n > 20
+    (Factorial for 21 will give you -> -4249290049419214848 which is incorrect)
+*/
+public class Factorial {
+    // Return n!
```
Comment (rohmulan) [inline Factorial/src/Factorial.java]: ? The name of function and class are descriptive enough, why this comment?

File: Factorial/src/Factorial.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    public static long factorialUsingRecursion(long n) {
+        if (n < 0) {
+            throw new RuntimeException("Underflow error in factorial");
```
Comment (rohmulan) [inline Factorial/src/Factorial.java]: Should be IllegalArgumentException

File: Factorial/src/Factorial.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        }
+        else if (n > 20) {
+            throw new RuntimeException("Overflow error in factorial");
```
Comment (rohmulan) [inline Factorial/src/Factorial.java]: Should be IllegalArgumentException

File: Factorial/src/Factorial.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            return 1;
+        } else {
+            return n * factorialUsingRecursion(n - 1);
+        }
+    }
+
+    @Test
+    public void factorialForAnyNumTest() {
+        long num = 4;
+        long actual = factorialUsingRecursion(num);
```
Comment (rohmulan) [inline Factorial/src/Factorial.java]: `long actual = factorialUsingRecursion(4);`
Why the unnecessary variable?

File: Factorial/src/Factorial.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        long expected = 24;
+        assertEquals(actual, expected);
```
Comment (rohmulan) [inline Factorial/src/Factorial.java:26]: Too many unnecessary variables. Just do: 
```java
assertEquals(factorialUsingRecursion(4), 24);
```

File: Factorial/src/Factorial.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Test
+    public void factorialForAnyNumGreaterThan20() {
+        long num = 22;
+        Exception exception = assertThrows(RuntimeException.class, () -> {
+            factorialUsingRecursion(num);
+        });
+        String expectedMessage = "Overflow error in factorial";
+        String actualMessage = exception.getMessage();
+        assertTrue(actualMessage.contains(expectedMessage));
```
Comment (rohmulan) [inline Factorial/src/Factorial.java:54]: This Assertion is weird. Why would you assert on error message like this? 
Use **assertThrows**

File: Factorial/src/Factorial.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            factorialUsingRecursion(num);
+        });
+        String expectedMessage = "Underflow error in factorial";
+        String actualMessage = exception.getMessage();
+        assertTrue(actualMessage.contains(expectedMessage));
+        System.out.println("factorialForAnyNumLessThan0 - passed");
+    }
+
+    @Test
+    public void factorialForAnyNumGreaterThan20() {
```
Comment (rohmulan) [inline Factorial/src/Factorial.java]: Give better test names. Test names should be like documentation of your function
```java
public void shouldThrowExceptionNumberGreaterThan20(){}
```

File: Factorial/src/Factorial.java (hunk 7 above)
Comment (ambicadeshpande) [reply Factorial/src/Factorial.java:26]: Isn't it good to have actual and expected variables while writing tests?
That is a nice to have practice usually? Correct me if I am wrong.
I would remove the num variable, yes that can be passed to a method directly

File: Factorial/src/Factorial.java (hunk 6 above)
Comment (ambicadeshpande) [reply Factorial/src/Factorial.java]: Similar reason for one of the comments you gave

File: Factorial/src/Factorial.java (hunk 3 above)
Comment (ambicadeshpande) [reply Factorial/src/Factorial.java]: Will remove those, had to post that on IG page.
Thanks.

File: Factorial/src/Factorial.java (hunk 2 above)
Comment (ambicadeshpande) [reply Factorial/src/Factorial.java]: Will remove them, had to post that on IG page.
Thanks.

File: Factorial/src/Factorial.java (hunk 1 above)
Comment (ambicadeshpande) [reply Factorial/src/Factorial.java]: Correct, will remove it

File: Factorial/src/Factorial.java (hunk 9 above)
Comment (ambicadeshpande) [reply Factorial/src/Factorial.java]: Makes sense

File: Factorial/src/Factorial.java (hunk 8 above)
Comment (ambicadeshpande) [reply Factorial/src/Factorial.java:54]: I did use **assetThrows** on line 62
This assertTrue is to verify the string that get returned when the exception is encountered

File: Factorial/src/Factorial.java (hunk 4 above)
Comment (ambicadeshpande) [reply Factorial/src/Factorial.java]: IllegalArgumentException extends the RuntimeException 
Ideally the greater than 20 value is not illegal? Just that the long data type will not be able to hold that size after computation right?

Hence did not add IllegalArgumentException then. But for the code it makes more sense to have this. Hence updating it.

**Pull request merged**